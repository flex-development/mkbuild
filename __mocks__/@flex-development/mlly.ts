/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
