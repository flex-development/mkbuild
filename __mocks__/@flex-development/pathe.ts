/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
