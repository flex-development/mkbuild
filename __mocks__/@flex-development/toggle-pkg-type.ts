/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
