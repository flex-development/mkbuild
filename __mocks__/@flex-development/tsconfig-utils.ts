/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
