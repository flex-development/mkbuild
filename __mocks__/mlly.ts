/**
 * @file Mocks - mlly
 * @module mocks/mlly
 * @see https://github.com/unjs/mlly
 */

/**
 * [`mlly`][1] module type.
 *
 * [1]: https://github.com/unjs/mlly
 */
type Actual = typeof import('mlly')

/**
 * `mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('mlly')

export const findDynamicImports = vi.fn(actual.findDynamicImports)
export const findExports = vi.fn(actual.findExports)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const resolvePath = vi.fn(actual.resolvePath)
