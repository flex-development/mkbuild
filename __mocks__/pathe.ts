/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
