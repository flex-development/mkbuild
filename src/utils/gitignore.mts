/**
 * @file Utilities - gitignore
 * @module utils/gitignore
 */

import dfs from '#internal/fs'
import toPath from '#internal/to-path'
import type { FileSystem } from '@flex-development/mkbuild'
import { isFile } from '@flex-development/mlly'
import pathe from '@flex-development/pathe'
import { ok } from 'devlop'

/**
 * Extract a list of ignore patterns from a `.gitignore` file.
 *
 * @see {@linkcode FileSystem}
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {URL | string | null | undefined} [dir]
 *  Module id of directory containing `.gitignore` file
 * @param {Partial<FileSystem> | null | undefined} [fs]
 *  File system adapter
 * @return {Set<string>}
 *  List of ignore patterns
 */
function gitignore(
  this: void,
  dir?: URL | string | null | undefined,
  fs?: Partial<FileSystem> | null | undefined
): Set<string> {
  fs = { ...dfs, ...fs }

  ok(typeof fs.readFileSync === 'function', 'expected `fs.readFileSync`')
  ok(typeof fs.statSync === 'function', 'expected `fs.statSync`')

  /**
   * Ignore patterns found in `.gitignore` file.
   *
   * @const {Set<string>} ignore
   */
  const ignore: Set<string> = new Set<string>()

  /**
   * Absolute path to `.gitignore` file.
   *
   * @const {string} path
   */
  const path: string = pathe.join(toPath(dir ?? pathe.cwd()), '.gitignore')

  if (isFile(path, fs as FileSystem)) {
    for (let line of fs.readFileSync(path, 'utf8').split(/\r?\n/)) {
      line = line.trim()

      if (!line || line.startsWith('#') || line.startsWith('!')) continue
      if (line.endsWith(pathe.sep)) line += '*'.repeat(2)
      ignore.add(line)
    }
  }

  return ignore
}

export default gitignore
