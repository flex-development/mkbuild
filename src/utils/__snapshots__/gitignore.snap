// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/gitignore > should return list of ignore patterns 1`] = `
Set {
  "**/._*",
  "**/.DS_*",
  "**/node_modules/**",
  "*.log",
  "*.pid",
  "*.pid.lock",
  "*.seed",
  "*.tgz",
  "*.zip",
  ".npm",
  ".yarn-integrity",
  "logs",
  "npm-debug.log*",
  "pids",
  "report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json",
  "yarn-debug.log*",
  "yarn-error.log*",
  ".pnp.*",
  ".yarn/*",
  "**/.env",
  "**/.env*.local",
  "**/.env.repo",
  "*.tsbuildinfo",
  "**/.vitest-reports/**",
  "**/__tests__/reports/*",
  "**/coverage/**",
  "**/tsconfig*temp.json",
  "*.lcov",
  "codecov",
  "**/.eslintcache",
  "**/dist/**",
  "**/*config.*.timestamp*",
  "**/.temp/**",
}
`;
